File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/code/CodeByRegionTypeLabelLanguageFilter.java
package org.taktik.icure.services.external.rest.v1.dto.filter.code;

import org.taktik.icure.entities.Patient;
import org.taktik.icure.entities.base.Code;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

import java.util.Objects;
import java.util.Optional;

import static org.taktik.icure.db.StringUtils.sanitizeString;

@JsonPolymorphismRoot(Filter.class)
public class CodeByRegionTypeLabelLanguageFilter extends Filter<Code> implements org.taktik.icure.dto.filter.code.CodeByRegionTypeLabelLanguageFilter {

    private String region;
    private String type;
    private String language;
    private String label;

    public String getRegion() {
        return region;
    }

    public void setRegion(String region) {
        this.region = region;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public String getLanguage() {
        return language;
    }

    public void setLanguage(String language) {
        this.language = language;
    }

    public String getLabel() {
        return label;
    }

    public void setLabel(String label) {
        this.label = label;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof CodeByRegionTypeLabelLanguageFilter)) return false;
        CodeByRegionTypeLabelLanguageFilter that = (CodeByRegionTypeLabelLanguageFilter) o;
        return Objects.equals(region, that.region) &&
                Objects.equals(type, that.type) &&
                Objects.equals(language, that.language) &&
                Objects.equals(label, that.label);
    }

    @Override
    public int hashCode() {
        return Objects.hash(region, type, language, label);
    }

    @Override
    public boolean matches(Code item) {
        String ss = sanitizeString(label);
        return (region == null || item.getRegions() == null || item.getRegions().contains(region))
                && (type == null || Objects.equals(type, this.type))
                && (language == null ? item.getLabel().values().stream().anyMatch( l -> Optional.ofNullable(l).map(s -> sanitizeString(s).contains(ss)).orElse(false))
                    : Optional.ofNullable(item.getLabel().get(language)).map(s -> sanitizeString(s).contains(ss)).orElse(false));
    }
}

File: src/main/kotlin/org/taktik/icure/domain/filter/impl/code/CodeByRegionTypeLabelLanguageFilter.kt
package org.taktik.icure.domain.filter.impl.code

import com.github.pozo.KotlinBuilder
import org.taktik.icure.db.StringUtils.sanitizeString
import org.taktik.icure.domain.filter.AbstractFilter
import org.taktik.icure.entities.base.Code
import java.util.*

@KotlinBuilder
data class CodeByRegionTypeLabelLanguageFilter(
        override val desc: String? = null,
        override val region: String? = null,
        override val type: String? = null,
        override val language: String? = null,
        override val label: String? = null
) : AbstractFilter<Code>, org.taktik.icure.domain.filter.code.CodeByRegionTypeLabelLanguageFilter {

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o !is CodeByRegionTypeLabelLanguageFilter) return false
        val that = o
        return region == that.region &&
                type == that.type &&
                language == that.language &&
                label == that.label
    }

    override fun hashCode(): Int {
        return Objects.hash(region, type, language, label)
    }

    override fun matches(item: Code): Boolean {
        val ss = sanitizeString(label)
        return ((region == null || item.regions.contains(region))
                && (type == null || type == type)
                && if (language == null) item.label.values.stream().anyMatch { l: String? -> Optional.ofNullable(l).map { s: String? -> sanitizeString(s)!!.contains(ss!!) }.orElse(false) } else Optional.ofNullable(item.label[language!!]).map { s: String? -> sanitizeString(s)!!.contains(ss!!) }.orElse(false))
    }
}
